analysis/src/machine_check.rs

```diff
@@ -13,31 +16,80 @@ use powdr_ast::{
         self,
         asm::{
             self, ASMModule, ASMProgram, AbsoluteSymbolPath, AssignmentRegister, FunctionStatement,
-            Instruction, LinkDeclaration, MachineProperties, MachineStatement, ModuleStatement,
-            RegisterFlag, SymbolDefinition,
+            Instruction, LinkDeclaration, MachineProperties, MachineStatement, Module,
+            ModuleStatement, RegisterFlag, SymbolDefinition, SymbolValue,
         },
+        EnumDeclaration, TraitDeclaration,
     },
 };
 
 /// Verifies certain properties of each machine and constructs the Machine objects.
 /// Also transfers generic PIL definitions but does not verify anything about them.
 pub fn check(file: ASMProgram) -> Result<AnalysisASMFile, Vec<String>> {
     let ctx = AbsoluteSymbolPath::default();
-    let machines = TypeChecker::default().check_module(file.main, &ctx)?;
+    let checker = TypeChecker::new(file.main);
+    checker.check_path(&ctx)?;
     Ok(AnalysisASMFile {
-        items: machines.into_iter().collect(),
+        items: checker.analyzed.into_inner(),
     })
 }
 
-#[derive(Default)]
-struct TypeChecker {}
+struct TypeChecker {
+    /// The parsed AST as a single [SymbolValue::Module]
+    parsed: SymbolValue,
+    /// The already checked paths
+    checked: RefCell<HashSet<AbsoluteSymbolPath>>,
+    /// The checked items
+    /// We use [RefCell] because we want to mutate this while borrowing from `parsed`
+    analyzed: RefCell<BTreeMap<AbsoluteSymbolPath, Item>>,
+}
 
 impl TypeChecker {
+    fn new(root: ASMModule) -> Self {
+        Self {
+            parsed: SymbolValue::Module(Module::Local(root)),
+            checked: Default::default(),
+            analyzed: Default::default(),
+        }
+    }
+
+    fn check_path(&self, ctx: &AbsoluteSymbolPath) -> Result<(), Vec<String>> {
+        // check if the path is already checked
+        if self.checked.borrow().contains(ctx) {
+            return Ok(());
+        }
+
+        let symbol = ctx.parts().fold(&self.parsed, |acc, part| match acc {
```

Comment:
This is quite expensive if there are many statements in the module. Why don't we have a map from names to items at this point? This is done after the path canonicalizer, right?

Reply:
Ah ok machine_check is the component that creates that map. Would it make sense to create the map first instead of iterating through the ast all the time here?

Reply:
Good point. It seems like the path canonicalizer should be the one responsible for creating the map (as it's the one making sure no key is duplicated). Let me try to implement that in a separate PR before this.

Reply:
Done in #1725. If it's just performance, we could merge this first and the other one later, up to you.